### PR TITLE
Improve jsonpath support in json functions

### PR DESCRIPTION
**Pull request by lgbo-ustc**
### Changelog category (leave one):

- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

support following new jsonpath format
- '$.1key', path element begins with number
- '$[key]', '$[“key”]', '$[\\\'key\\\']', '$["key 123"]', path element embraced in []

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp [hunk 1]
```diff
@@ -16,18 +18,48 @@ namespace DB
  */
 bool ParserJSONPathMemberAccess::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
 {
-    if (pos->type != TokenType::Dot)
+    // There's a special case, that a path member can begin with number
+    if (pos->type != TokenType::Dot && pos->type != TokenType::Number)
         return false;
+    if (pos->type != TokenType::Number)
+        ++pos;
```
Comment (kssenii) [inline src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp:25]: did not understand why do we need this condition `if (pos->type != TokenType::Number)` and same above if we expect number after "." and here we are supposed to skip "." and only after comes a number, from your own example in test: `'$.1key'`

Comment (lgbo-ustc) [reply src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp:25]: ".1123" is parsed as a number

Comment (lgbo-ustc) [reply src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp:25]: ".123" is parsed as a number, not a dot and a number

Comment (lgbo-ustc) [reply src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp:25]: I'm not sure whether change the lexer will bring some compatibility issues. So make the change here will be safty.

File: src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp [hunk 2]
```diff
@@ -16,18 +18,48 @@ namespace DB
...
 
-    if (pos->type != TokenType::BareWord && pos->type !=TokenType::QuotedIdentifier)
-        return false;
+    if (pos->type == TokenType::Number)[[unlikely]]
+    {
+        for (const auto * c = pos->begin; c != pos->end; ++c)
+        {
+            if (*c == '.' && c == pos->begin)
+                continue;
+            if (!isNumericASCII(*c))
```
Comment (kssenii) [inline src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp:35]: as I understand here if number does not start with `.` then we have something like `$1abc`, which is not valid, but here it seems we allow such case? let's add a test and make sure it is not parsed

File: src/Functions/JSONPath/Parsers/ParserJSONPathMemberSquareBracketAccess.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            ReadBufferFromMemory in(pos->begin, pos->size());
+            String name;
+            readQuotedStringWithSQLStyle(name, in);
+            member_name = std::make_shared<ASTIdentifier>(name);
+            ++pos;
+        }
+        catch (const Exception &)
+        {
+            return false;
```
Comment (kssenii) [inline src/Functions/JSONPath/Parsers/ParserJSONPathMemberSquareBracketAccess.cpp:35]: a comment is needed, with explanation why we need try catch.

File: src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp (hunk 1 above)
Comment (kssenii) [reply src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp:25]: please put all this into comment in the code

File: src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp (hunk 2 above)
Comment (lgbo-ustc) [reply src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp:35]: I think `$` always follow by a dot or a square bracket

File: src/Functions/JSONPath/Parsers/ParserJSONPathMemberSquareBracketAccess.cpp (hunk 3 above)
Comment (lgbo-ustc) [reply src/Functions/JSONPath/Parsers/ParserJSONPathMemberSquareBracketAccess.cpp:35]: It's for early debug. I remove it.

File: src/Functions/JSONPath/Parsers/ParserJSONPathMemberSquareBracketAccess.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    ++pos;
+    ASTPtr member_name;
+    if (pos->type == TokenType::BareWord || pos->type == TokenType::QuotedIdentifier)
+    {
+        ParserIdentifier name_p;
+        if (!name_p.parse(pos, member_name, expected))
+            return false;
+    }
+    else if (pos->type == TokenType::StringLiteral)
+    {
```
[the remark this answers is not in the document]
Comment (lgbo-ustc) [reply src/Functions/JSONPath/Parsers/ParserJSONPathMemberSquareBracketAccess.cpp:24]: If you do like this, you still have to make an condition `pos->type == TokenType::StringLiteral` for parsing `name` from   StringLiteral

File: src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp (hunk 2 above)
Comment (kssenii) [reply src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp:51]: > I think $ is always followed by a dot or a square bracket

this point is - it should be, but we need to make sure that if not - then this case returns a "cannot parse" error. So I suggested to add a test for this case.

File: src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp [hunk 5]
```diff
@@ -16,18 +20,62 @@ namespace DB
  */
 bool ParserJSONPathMemberAccess::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
 {
-    if (pos->type != TokenType::Dot)
+    LOG_ERROR(&Poco::Logger::get("ParserJSONPathMemberAccess"), "xxx {} pos type: {}. token:{}", __LINE__, pos->type, std::string_view(pos->begin, pos->size()));
```
Comment (zhanglistar) [inline src/Functions/JSONPath/Parsers/ParserJSONPathMemberAccess.cpp:23]: remember to del.

**Review by kssenii — APPROVED** [review]

**Pull request merged**